eth_p2p/rlpx_protocols/shh_protocol.nim

```diff
@@ -416,31 +539,406 @@ proc add*(self: var Queue, msg: Message) =
+  # NOTE: Should we allow a filter without a key? Encryption is mandatory in v6?
+  # Check if asymmetric _and_ symmetric key? Now asymmetric just has precedence.
+  let id = generateRandomID()
+  var filter = filter
+  if handler.isSome():
+   filter.handler = handler
+  else:
+   filter.queue = newSeqOfCap[ReceivedMessage](defaultFilterQueueCapacity)
+
+  filters.add(id, filter)
+  debug "Filter added", filter = id
+  return id
+
+proc notify*(filters: var Filters, msg: Message) =
+ var decoded: Option[DecodedPayload]
+ var keyHash: Hash
+
+ for filter in filters.mvalues:
+   if not filter.allowP2P and msg.isP2P:
+     continue
+
+   # if message is direct p2p PoW doesn't matter
+   if msg.pow < filter.powReq and not msg.isP2P:
+     continue
+
+   if filter.topics.len > 0:
+     if msg.env.topic notin filter.topics:
+       continue
+
+   # Decode, if already decoded previously check if hash of key matches
+   if decoded.isNone():
+     decoded = decode(msg.env.data, dst = filter.privateKey,
+                      symKey = filter.symKey)
+     if filter.privateKey.isSome():
+       keyHash = keccak256.digest(filter.privateKey.get().data)
+     elif filter.symKey.isSome():
+       keyHash = keccak256.digest(filter.symKey.get())
+     # else:
+       # NOTE: should we error on messages without encryption?
+     if decoded.isNone():
+       continue
+   else:
+     if filter.privateKey.isSome():
+       if keyHash != keccak256.digest(filter.privateKey.get().data):
+         continue
+     elif filter.symKey.isSome():
+       if keyHash != keccak256.digest(filter.symKey.get()):
+         continue
+     # else:
+       # NOTE: should we error on messages without encryption?
+
+   # When decoding is done we can check the src (signature)
+   if filter.src.isSome():
+     let src: Option[PublicKey] = decoded.get().src
+     if not src.isSome():
+       continue
+     elif src.get() != filter.src.get():
+       continue
+
+   let receivedMsg = ReceivedMessage(decoded: decoded.get(),
+                                     timestamp: msg.env.expiry - msg.env.ttl,
+                                     ttl: msg.env.ttl,
+                                     topic: msg.env.topic,
+                                     pow: msg.pow,
+                                     hash: msg.hash)
+   # Either run callback or add to queue
+   if filter.handler.isSome():
+     filter.handler.get()(receivedMsg)
+   else:
+     filter.queue.insert(receivedMsg)
+
+proc getFilterMessages*(filters: var Filters, filterId: string): seq[ReceivedMessage] =
+  result = @[]
+  if filters.contains(filterId):
+    if filters[filterId].handler.isNone():
+      result = filters[filterId].queue
+      filters[filterId].queue =
+        newSeqOfCap[ReceivedMessage](defaultFilterQueueCapacity)
+
+proc toBloom*(filters: Filters): Bloom =
+  for filter in filters.values:
+    if filter.topics.len > 0:
+      result = result or filter.bloom
+
+type
+  PeerState = ref object
```

Comment:
I'll rename these to make them a bit more consistent with the other modules (I'll push a commit). The new names will be `WhisperPeer` and `WhisperNetwork`

Reply:
OK, sounds good